iwyu.cc

```diff
@@ -3115,45 +3128,126 @@ class InstantiatedTemplateVisitor
     // to require it as well.  TODO(csilvers): this doesn't really
     // make any sense.  Who figures out we need the full type if
     // you do 'Foo<MyClass>::value_type m;'?
-    for (const ASTNode* ast_node = current_ast_node();
+    for (const ASTNode* ast_node = node;
          ast_node != caller_ast_node_; ast_node = ast_node->parent()) {
-      if (ast_node->IsA<TypedefNameDecl>())
-        return Base::VisitSubstTemplateTypeParmType(type);
+      if (ast_node->IsA<TypedefNameDecl>()) {
+        return false;
+      }
+      if (ast_node->IsA<TemplateSpecializationType>()) {
+        // If we hit a template specialization node before the typedef then we
+        // probably still need a full-use, so stop looking.
+        break;
+      }
     }
 
     // sizeof(a reference type) is the same as sizeof(underlying type).
     // We have to handle that specially here, or else we'll say the
     // reference is forward-declarable, below.
-    if (current_ast_node()->ParentIsA<UnaryExprOrTypeTraitExpr>() &&
-        isa<ReferenceType>(actual_type)) {
-      const ReferenceType* actual_reftype = cast<ReferenceType>(actual_type);
-      ReportTypeUse(CurrentLoc(),
-                    actual_reftype->getPointeeTypeAsWritten().getTypePtr());
-      return Base::VisitSubstTemplateTypeParmType(type);
+    if (node->ParentIsA<UnaryExprOrTypeTraitExpr>() &&
+        isa<ReferenceType>(type)) {
+      return true;
     }
 
     // If we're used in a forward-declare context (MyFunc<T>() { T* t; }),
     // or are ourselves a pointer type (MyFunc<Myclass*>()),
     // we don't need to do anything: we're fine being forward-declared.
-    if (current_ast_node()->in_forward_declare_context())
-      return Base::VisitSubstTemplateTypeParmType(type);
+    if (node->in_forward_declare_context())
+      return false;
 
-    if (current_ast_node()->ParentIsA<PointerType>() ||
-        current_ast_node()->ParentIsA<LValueReferenceType>() ||
-        IsPointerOrReferenceAsWritten(actual_type))
-      return Base::VisitSubstTemplateTypeParmType(type);
+    if (node->ParentIsA<PointerType>() ||
+        node->ParentIsA<LValueReferenceType>() ||
+        IsPointerOrReferenceAsWritten(type))
+      return false;
 
-    // We attribute all uses in an instantiated template to the
-    // template's caller.
-    ReportTypeUse(caller_loc(), actual_type);
+    return true;
+  }
+
+  // This helper is called on every use of a template argument type in an
+  // instantiated template.  Its goal is to determine whether that use should
+  // constitute a full-use by the template caller, and perform other necessary
+  // bookkeeping.
+  void AnalyzeTemplateTypeParmUse(const Type* type)
```

Comment:
How about just `ReportTemplateTypeParmUse`?

Hmm, I see it's also registering `template_argument_aliases_`, I guess that's the "other necessary bookkeeping".

I wonder if you could make that more explicit by e.g. having callers pass in the `template_argument_aliases_` container? On balance, I'm not sure if it's an improvement, but worth playing around with.

Reply:
I don't think it helps.  Etiehr way it's just member functions using a member variable.  And moving it to the callers means one more reference to that variable.